### ISSUE TITLE
<a-Z> should always prompt for a combining operator.

### ISSUE THREAD

**Issue by Screwtapello** [mawww/kakoune#3374]
### Steps 

1. Start Kakoune without any scripts:

        kak -n

1. Save the selection to the `caret` register:

        Z

1. Move the cursor, then append the new selection to the `caret` register:

        l<a-Z>a

1. Clear the `caret` register:

        :set-register caret<ret>

1. Try to append the new selection to the now-empty `caret` register:

        <a-Z>

### Outcome 

Kakoune displays the message:

> Combined 1 selections to register '^'

### Expected 

The "combine selections to register" menu should appear, asking how the selections should be combined, just as it was when the register wasn't empty.

### Notes

If the register is empty, there's only one way to combine the current selections (append), so strictly speaking the menu isn't necessary. However, this makes it more difficult for a script to build up a set of selections.

Let's say a script wanted to build up a list of all the octal integer literals in the selection, using both the traditional (`0755`) and modern (`0o755`) syntaxes. If Kakoune always prompted for a combining operator, the script might look something like this.

```kak
eval -save-regs ab %{
    # Store the original selection in a
    exec '"a' Z

    # Gather traditional octal literals
    try %{
        # Restore the original selection
        exec '"a" z

        # Find any traditional octal literals
        exec s0[0-7]+<ret>

        # Store them in "b
        exec '"b' <a-Z>a
    }

    # Gather modern octal literals
    try %{
        # Restore the original selection
        exec '"a" z

        # Find any traditional octal literals
        exec s0o[0-7]+<ret>

        # Store them in "b
        exec '"b' <a-Z>a
    }

    # Now "b contains the coordinates of all the octal literals
}
```

Under the current scheme, this doesn't work reliably: if the selection contains traditional octal literals, the modern-literals `try` block needs to use `<a-Z>a`, but otherwise the modern-literals block needs to use `<a-Z>` with out the following `a`. It's possible to write a shell-script that emits `<a-Z>` or `<a-Z>a` after inspecting the current register contents, but it would be much easier to make Kakoune just ask every time.

On IRC, mawww said:

> I think its a historical baggage of what <a-Z> did before (appending  instead of combining)
>
> But it would seems better to handle the empty case as part of the combination operation
>
> Especially as selection combination operators are mostly used in scripting, so they really need to be predictible

As a clue to implementation, try removing the `empty` variable and the code that checks it from [the `save_selections()` function](https://github.com/mawww/kakoune/blob/61cae516054aa4cb0733f37bf26636a950f7999b/src/normal.cc#L1894-L1920).

**Comment by lenormf** [comment]
Is this related to #3315? Dupe?

**Comment by Screwtapello** [comment]
Yeah, it does seem to be a dupe of that issue, although this one has more detail.

**Comment by nonumeros** [comment]
I don't get it at all. I'm deeply sorry. 

But if you clear the register, how on earth is `combined 1 selection to register '^'` not supposed to appear? 

You said:

> Expected
The "combine selections to register" menu should appear, asking how the selections should be combined, just as it was when the register wasn't empty.

How is it going to show up at all when it was 'cleared'? It's the case presumably that it's empty....

How is it going to keep it up combining it for crying out loud...

**Comment by Screwtapello** [comment]
The menu that (normally) appears with `<a-Z>` contains the following items:

- append
- union
- intersection
- select leftmost cursor
- select rightmost cursor
- select longest
- select shortest

If the buffer has two selections but the register contains one selection, "append" will just give you three selections in the register. On the other hand, If the register is empty, "append" gives you two selections in the register - "append" works the same way no matter how many selections are in the register.

Again, if the buffer has two selections but the register contains one, all the other operations in the menu will fail, because they try to pair up each buffer selection with a register selection. If the register is empty, they *still* fail in exactly the same way, and for exactly the same reason. Again, these options work the same way no matter how many selections are in the register.

Therefore, if all the items in the menu work the same way whether the register is empty or not, it doesn't make sense to skip the menu in the "empty register" case.

**Comment by nonumeros** [comment]
@Screwtapello  thank you for the reminder with the menu contents.

In all honesty I don't know if there are other underlying problems of which I'm totally clueless about. But I'm fine with it. 

With the related issue that you linked from @alexherbo2, I'm still having the same problems than before, and I don't know if it's a feature or an error. At this point I don't even know what a feature is and  when one is not.    

> If the buffer has two selections but the register contains one selection, "append" will just give you three selections in the register. On the other hand, If the register is empty, "append" gives you two selections in the register - "append" works the same way no matter how many selections are in the register.

I agree.

But back to this issue:

If I were to combine only one (1) selection to register and then I were to clear that very same register, how am I supposed to have another selection.
 
In other words, yes, `Z`, `<a-Z>al` it's nonetheless always to but never  from that register.

Then you clear the register which is undoing what was previously done.

Had another selection were to be appended through the menu then I think <a-z> which combines two (2) selections rather than `<a-Z>` that's not doing anything else, would suffice.

Can the same behavior be reproduced by replacing the steps with `<a-z>`? Im just  guessing  because no matter the number of selections combined to the register, the register will be cleared. 

```
Z l <a-z>al
<a-z>al<a-z>al
<a-Z>al 
:set-register caret<ret>
<a-Z>al
<a-Z>al
```

should return the correct number of selections.

**Comment by Screwtapello** [comment]
I guess you could make the argument that `<a-z>` should always display a menu and `<a-z>a` should work even if the register is empty, but I haven't needed that behaviour yet, so I didn't mention it.
